cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetInstanceMDSUpgrader.java

```diff
@@ -0,0 +1,155 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.datafabric.dataset.service.mds;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.DatasetDefinition;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.datafabric.dataset.DatasetMetaTableUtil;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.data2.dataset2.tx.Transactional;
+import co.cask.cdap.proto.Id;
+import co.cask.tephra.TransactionExecutor;
+import co.cask.tephra.TransactionExecutorFactory;
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Dataset instances metadata store
+ */
+public final class DatasetInstanceMDSUpgrader {
+  private static final Logger LOG = LoggerFactory.getLogger(DatasetInstanceMDSUpgrader.class);
+  private final Transactional<UpgradeMdsStores<DatasetInstanceMDS>, DatasetInstanceMDS> datasetInstanceMds;
+
+  @Inject
+  private DatasetInstanceMDSUpgrader(final TransactionExecutorFactory executorFactory,
+                                     @Named("dsFramework") final DatasetFramework dsFramework) {
+
+    this.datasetInstanceMds = Transactional.of(executorFactory,
+       new Supplier<UpgradeMdsStores<DatasetInstanceMDS>>() {
+         @Override
+         public UpgradeMdsStores<DatasetInstanceMDS> get() {
+           String dsName = Joiner.on(".").join(Constants.SYSTEM_NAMESPACE, DatasetMetaTableUtil.INSTANCE_TABLE_NAME);
+           Id.DatasetInstance datasetId = Id.DatasetInstance.from(Constants.DEFAULT_NAMESPACE_ID, dsName);
+           try {
+             DatasetInstanceMDS oldMds =
+               DatasetsUtil.getOrCreateDataset(dsFramework, datasetId, DatasetInstanceMDS.class.getName(),
+                                               DatasetProperties.EMPTY, DatasetDefinition.NO_ARGUMENTS, null);
+             DatasetInstanceMDS newMds = new DatasetMetaTableUtil(dsFramework).getInstanceMetaTable();
+             return new UpgradeMdsStores<DatasetInstanceMDS>(oldMds, newMds);
+           } catch (Exception e) {
+             LOG.error("Failed to access table: {}", datasetId, e);
+             throw Throwables.propagate(e);
```

Comment:
ouch. Can we have some class for UpgradeFailedException or smth - smth "checked", so that upgrade tool itself could act on it appropriately? catch - revert - tell user smth - whatever